apisix/plugins/ldap-auth.lua

```diff
@@ -0,0 +1,171 @@
+        uid = { type = "string" },
+        auto_create_consumer = {type ="boolean"}
+    },
+    required = {"basedn","ldapuri"},
+    additionalProperties = true,
+}
+
+local consumer_schema = {
+    type = "object",
+    title = "work with consumer object",
+    properties = {
+        userdn = { type = "string" },
+    },
+    required = {"userdn"},
+    additionalProperties = false,
+}
+
+local plugin_name = "ldap-auth"
+
+local _M = {
+    version = 0.1,
+    priority = 2520,
+    type = 'auth',
+    name = plugin_name,
+    schema = schema,
+    consumer_schema = consumer_schema
+}
+
+function _M.check_schema(conf, schema_type)
+    local ok, err
+    if schema_type == core.schema.TYPE_CONSUMER then
+        ok, err = core.schema.check(consumer_schema, conf)
+    else
+        ok, err = core.schema.check(schema, conf)
+    end
+
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+local create_consume_cache
+do
+    local consumer_names = {}
+
+    function create_consume_cache(consumers)
+        core.table.clear(consumer_names)
+
+        for _, consumer in ipairs(consumers.nodes) do
+            core.log.info("consumer node: ", core.json.delay_encode(consumer))
+            consumer_names[consumer.auth_conf.userdn] = consumer
+        end
+
+        return consumer_names
+    end
+
+end -- do
+
+local function extract_auth_header(authorization)
+    local obj = { username = "", password = "" }
+
+    local m, err = ngx.re.match(authorization, "Basic\\s(.+)", "jo")
+    if err then
+        -- error authorization
+        return nil, err
+    end
+
+    local decoded = ngx.decode_base64(m[1])
+
+    local res
+    res, err = ngx_re.split(decoded, ":")
+    if err then
+        return nil, "split authorization err:" .. err
+    end
+
+    obj.username = ngx.re.gsub(res[1], "\\s+", "", "jo")
+    obj.password = ngx.re.gsub(res[2], "\\s+", "", "jo")
+
+    return obj, nil
+end
+
+function _M.rewrite(conf, ctx)
+    core.log.info("plugin access phase, conf: ", core.json.delay_encode(conf))
+
+    -- 1. extract authorization from header
+    local auth_header = core.request.header(ctx, "Authorization")
+    if not auth_header then
+        core.response.set_header("WWW-Authenticate", "Basic realm='.'")
+        return 401, { message = "Missing authorization in request" }
+    end
+
+    local user, err = extract_auth_header(auth_header)
+    if err then
+        return 401, { message = err }
+    end
+
+    -- 2. try authenticate the user against the ldap server
+    local uid = "cn"
+    if conf.uid then 
+        uid = conf.uid
+    end
+    local userdn =  uid .. "=" .. user.username .. "," .. conf.basedn
+    local ld = lualdap.open_simple (conf.ldapuri, userdn, user.password, conf.usetls)
+    if not ld then
+        return 401, { message = "Invalid user authorization" }
+    end
+    
+    -- 3. Retreive consumer for authorization plugin
+    if conf.auto_create_consumer then 
```

Comment:
Better to avoid creating temporary consumers. The temporary consumer may be broken somewhere else.

Reply:
Ok remove this feature .
It was more to act like the oidc plugin , that does not require a consumer to be created , but i see your point